src/api-service/__app__/timer_retention/__init__.py

```diff
@@ -0,0 +1,97 @@
+import datetime
+import logging
+
+import azure.functions as func
+from onefuzztypes.enums import JobState, TaskState
+
+from ..onefuzzlib.events import get_events
+from ..onefuzzlib.jobs import Job
+from ..onefuzzlib.notifications.main import Notification
+from ..onefuzzlib.repro import Repro
+from ..onefuzzlib.tasks.main import Task
+
+RETENTION_POLICY = datetime.timedelta(minutes=(5))
+
+
+def main(mytimer1: func.TimerRequest, dashboard: func.Out[str]) -> None:  # noqa: F841
+
+    now = datetime.datetime.now(tz=datetime.timezone.utc)
+
+    time_retained = now - RETENTION_POLICY
+
+    time_filter = f"Timestamp lt datetime'{time_retained.isoformat()}'"
+
+    # You have to do notification before task,
+    # because editing the upn for tasks will change the timestamp
+    for notification in Notification.search(raw_unchecked_filter=time_filter):
+        logging.info("Retention Timer Notification Search")
+        logging.info(
+            "Found notification %s older than 18 months. Checking related tasks.",
+            notification.notification_id,
+        )
+        container = notification.container
+        timestamp_list = []
+        for task in Task.search():

```

Comment:
It's my understanding, you want the logic to be:  "delete notifications older than X where no tasks were used since Y".

You should be able to query tasks once, rather than each time you query the notification.  Additionally, you should filter the tasks to only query tasks within the window you care about it.

Reply:
Okay so the reason I do a blanket search on tasks each time is because I don't want to only query Task.search("older than 18 months") if there are potentially tasks that are newer that still reference the notification/container. 

I.e., Let's say I have a notification that references container-123. I have a set of tasks older than 18 months and a set of tasks younger than 18 months. If I only query Task.search("older than 18 months"), I might inappropriately delete the notification. 

Unless, of course, I'm misunderstanding you're feedback and you're suggesting I flip the loops such that I iterate over the tasks and _then_ iterate over notifications?

Reply:
Right now, you are calling `Task.search` for every notification.

If you have 300 notifications, you are calling `Task.search` 300 times.  Task.search, as shown here, returns all tasks that have ever existed.

I'm suggesting changing the process to the following:

Get the containers used by all tasks newer than 18 months, call this Set A.  (Only do this once, not once per notification).
For each notification older than 18 months, if the notification.container is not in Set A, clean up the notification.

Reply:
AH!

Excellent okay I understand. Thanks Brian.